File: Video_ZeromqPublisher.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <zmq.hpp>

int main() 
{
  // Initialize ZeroMQ context
  zmq::context_t context(1);
  zmq::socket_t publisher(context, ZMQ_PUB);
  publisher.bind("tcp://*:5555");

  // Open the camera
  cv::VideoCapture cap(0);
  if (!cap.isOpened()) 
  {
    std::cerr << "Error opening video stream or file" << std::endl;
    return -1;
  }

  while (true) 
  {
    cv::Mat frame;
    cap >> frame;

    // Serialize the frame to a binary message
    std::vector < uchar > buffer;
    cv::imencode(".jpg", frame, buffer);

    // Send the frame over ZeroMQ
    zmq::message_t message(buffer.size());
    memcpy(message.data(), buffer.data(), buffer.size());
    publisher.send(message, zmq::send_flags::none);
  }

  return 0;
}

File: Video_ZeromqPublisher.py

import cv2
import zmq

# Initialize OpenCV video capture
cap = cv2.VideoCapture(0)  # 0 for default camera

# Initialize ZeroMQ context and socket
context = zmq.Context()
socket = context.socket(zmq.PUB)
socket.bind("tcp://*:5555")  # Bind to a local port

while True:
    # Capture frame from OpenCV
    ret, frame = cap.read()

    # Encode the frame as JPEG
    _, buffer = cv2.imencode('.jpg', frame)

    # Send the frame over ZeroMQ
    socket.send(buffer.tobytes())

# Release resources
cap.release()
cv2.destroyAllWindows()
